python/ember/input.py

```diff
@@ -1314,6 +1342,183 @@ def run(self):
         solver.finalize()
         return solver
 
+    def runESR(self):
```

Comment:
In general, I think this method needs some additional comments which explain what is going on and why. The algorithm isn't that obvious without a bit of context.